### PR TITLE
Adding support for parallel execution of builders

### PR DESCRIPTION
**Pull request by ankurs**

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -16,10 +16,10 @@ jobs:
     - name: Set up Go
       uses: actions/setup-go@v2
       with:
-        go-version: 1.17
+        go-version: 1.18
```
Comment (peterjrichens) [inline .github/workflows/go.yml:19]: curious why you changed the version here when `go.mod` still references 1.17?

File: plan.go [hunk 2]
```diff
@@ -40,17 +46,23 @@ func (p *plan) Replace(ctx context.Context, from interface{}, to interface{}) er
...
 		}
 	}
 	return errors.New("builder not found")
 }
 
 func (p *plan) Run(ctx context.Context, initData ...interface{}) (Result, error) {
+	return p.RunParallel(ctx, 1, initData...)
+}
+
+func (p *plan) RunParallel(ctx context.Context, workers uint, initData ...interface{}) (Result, error) {
```
Comment (peterjrichens) [inline plan.go:66]: I'm wondering if a go routine leak is possible. I added `goleak.VerifyNone(t)` (using https://github.com/uber-go/goleak) to `TestRunPlan` and got this error:

```
    leaks.go:78: found unexpected goroutines:
        [Goroutine 30 in state chan receive, with github.com/go-coldbrew/data-builder.worker on top of the stack:
        goroutine 30 [chan receive]:
        github.com/go-coldbrew/data-builder.worker({0x47ac9d0, 0xc000016148}, 0xc0001b8680?)
        	/Users/peter/dev/data-builder/plan.go:103 +0x7d
        created by github.com/go-coldbrew/data-builder.(*plan).run
        	/Users/peter/dev/data-builder/plan.go:176 +0x9a
        ]
```

Comment (peterjrichens) [inline plan.go:66]: How feasible would it be to provide the user with a maxWorker option? ie. after resolving the dependencies can data-builder see the maximum amout of parallelism possible and create this many workers?

File: plan.go [hunk 3]
```diff
@@ -71,40 +83,108 @@ func (p *plan) Run(ctx context.Context, initData ...interface{}) (Result, error)
...
 	return nil
 }
 
+func (p *plan) run(ctx context.Context, workers uint, dataMap map[string]interface{}) error {
+	if workers == 0 {
+		workers = 1
+	}
+
+	// create a work channel and start workers
+	wChan := make(chan work, 0)
```
Comment (peterjrichens) [inline plan.go:175]: do we need to handle closing of this channel?

File: .github/workflows/go.yml (hunk 1 above)
Comment (ankurs) [reply .github/workflows/go.yml:19]: I just updated the gitlab runner version, older version of go.mod are still compatible with new versions :)

File: plan.go (hunk 3 above)
Comment (ankurs) [reply plan.go:175]: thats a bug haha, this is what is causing go-routine leak, let me fix it and add a check for go-routine

File: plan.go (hunk 2 above)
Comment (ankurs) [reply plan.go:66]: let me fix it

Comment (ankurs) [reply plan.go:66]: thats an interesting question, Normally I would suggest number of workers to be proportional to cpu cores/threads.
spawning more workers than number of cores/threads may not always be helpful, specially for cpu intensive tasks.

I can add an option to read max parallelism from a given plan, what interface will you prefer ?

Comment (peterjrichens) [reply plan.go:66]: I see, having the user restrict the number of workers makes sense in a cpu-intensive scenario. but in my use case, I know network calls will be the bottleneck. should we replace the `workers` arg with `maxWorkers`? the actual number of workers used could be `min(plan.maxParallelism(), maxWorkers)`. eg. suppose resolve dependencies returns:
```
{{"Name3"}, {"Name2", "Name5"}, {"Name1", "Name4"}}
```
in this case `plan.maxParallelism()` = 2 as we know there won't be any major advantage to setting workers > 2

Comment (ankurs) [reply plan.go:66]: yes, this should be simple enough to implement, do you think maybe I can create a static function in data-builder, something along the lines of
```go
func MaxPlanParallelism(p Plan) (uint, error)
{
    ....
}
```

Comment (ankurs) [reply plan.go:66]: I have added this function

**Review by peterjrichens — APPROVED** [review]
LGTM

**Pull request merged**